### PR TITLE
Add `ConvectionIntegrator::AddMultTransposePA`.

### PR DESCRIPTION
**Pull request by YohannDudouit**
Add `ConvectionIntegrator::AddMultTransposePA`.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#2840](https://github.com/mfem/mfem/pull/2840) | @YohannDudouit | @tzanio | @artv3 + @tzanio | 2/24/22 | 2/24/22 | 2/28/22 | |

### CODE REVIEW

**Comment by YohannDudouit** [comment]
@v-dobrev @pazner Any opinion how I should test this? Should I add a test with:
```c++
k.AddDomainIntegrator(
      new TransposeIntegrator(new ConvectionIntegrator(velocity, -1.0)));
```
?

**Comment by pazner** [comment]
It would probably be most straightforward to add such a test case to `test_pa_kernels.cpp` (e.g. as an option to `test_pa_convection`).

**Review by artv3 — APPROVED** [review]
Works for me!

File: fem/bilininteg_convection_pa.cpp [hunk 1]
```diff
@@ -766,6 +766,615 @@ void SmemPAConvectionApply3D(const int ne,
...
+   MFEM_FORALL_2D(e, NE, Q1D, Q1D, NBZ,
+   {
+      const int tidz = MFEM_THREAD_ID(z);
+      const int D1D = T_D1D ? T_D1D : d1d;
+      const int Q1D = T_Q1D ? T_Q1D : q1d;
+      // the following variables are evaluated at compile time
+      constexpr int NBZ = T_NBZ ? T_NBZ : 1;
+      constexpr int max_D1D = T_D1D ? T_D1D : MAX_D1D;
+      constexpr int max_Q1D = T_Q1D ? T_Q1D : MAX_Q1D;
+      // constexpr int MDQ = (max_Q1D > max_D1D) ? max_Q1D : max_D1D;
```
Comment (artv3) [inline fem/bilininteg_convection_pa.cpp:918]: Perhaps remove MDQ if not used.

**Comment by tzanio** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

**Review by tzanio — APPROVED** [review]

**Comment by tzanio** [comment]
Merged in `next` for testing...

File: fem/bilinearform.cpp [hunk 2]
```diff
@@ -1064,6 +1064,18 @@ void BilinearForm::Mult(const Vector &x, Vector &y) const
...
 
+void BilinearForm::MultTranspose(const Vector & x, Vector & y) const
+{
+   if (ext)
+   {
+      ext->MultTranspose(x, y);
+   }
+   else
+   {
+      y = 0.0; AddMultTranspose (x, y);
```
Comment (pazner) [inline fem/bilinearform.cpp:1075]:
```suggestion
      y = 0.0;
      AddMultTranspose (x, y);
```

File: tests/unit/fem/test_pa_kernels.cpp [hunk 3]
```diff
@@ -445,7 +454,33 @@ TEST_CASE("PA Convection", "[PartialAssembly][MFEMData]")
...
+      // - 0: CG,
+      // - 1: DG continuous coeff,
+      // - 2: DG discontinuous coeff,
+      // - 3: DG Bernstein discontinuous coeff.
+      auto prob = GENERATE(0, 1, 2, 3);
+      auto order_2d = GENERATE(2);
+      auto order_3d = GENERATE(2);
+      // refinement > 0 => Non-conforming mesh
+      auto refinement_2d = GENERATE(0,1);
+      auto refinement_3d = GENERATE(0,1);
```
Comment (pazner) [inline tests/unit/fem/test_pa_kernels.cpp:481]: You should put the 2D and 3D `GENERATE` inside the respective `SECTION`s to avoid running duplicate tests

Comment (pazner) [reply tests/unit/fem/test_pa_kernels.cpp:481]: Similarly for the `"PA Convection"` test case

File: fem/bilinearform_ext.cpp [hunk 4]
```diff
@@ -874,6 +878,9 @@ void FABilinearFormExtension::Assemble()
             static_cast<const L2ElementRestriction*>(elem_restrict);
          const L2FaceRestriction *restF =
             static_cast<const L2FaceRestriction*>(int_face_restrict_lex);
+         MFEM_VERIFY(
+            !dynamic_cast<const NCL2FaceRestriction*>(int_face_restrict_lex),
+            "Full Assembly not yet supported on NCMesh.");
```
Comment (pazner) [inline fem/bilinearform_ext.cpp:883]: Ditto

File: fem/bilinearform_ext.cpp [hunk 5]
```diff
@@ -545,6 +545,8 @@ void EABilinearFormExtension::Assemble()
                                                      ea_data_ext,
                                                      i);
    }
+   MFEM_VERIFY(!dynamic_cast<const NCL2FaceRestriction*>(int_face_restrict_lex),
+               "Element Assembly not yet supported on NCMesh.");
```
Comment (pazner) [inline fem/bilinearform_ext.cpp:549]: Not easier to just check e.g. `trial_fes->Conforming()`?

File: fem/bilinearform_ext.cpp [hunk 6]
```diff
@@ -558,16 +560,18 @@ void EABilinearFormExtension::Assemble()
    {
       bdrFaceIntegrators[i]->AssembleEABoundaryFaces(*a->FESpace(),ea_data_bdr,i);
    }
+   MFEM_VERIFY(!dynamic_cast<const NCL2FaceRestriction*>(bdr_face_restrict_lex),
+               "Element Assembly not yet supported on NCMesh.");
```
Comment (pazner) [inline fem/bilinearform_ext.cpp:564]: Ditto

File: fem/bilinearform_ext.cpp [hunk 7]
```diff
@@ -558,16 +560,18 @@ void EABilinearFormExtension::Assemble()
...
-      restFint.AddFaceMatricesToElementMatrices(ea_data_int, ea_data);
+      auto restFint = dynamic_cast<const L2FaceRestriction*>(int_face_restrict_lex);
+      restFint->AddFaceMatricesToElementMatrices(ea_data_int, ea_data);
    }
    if (factorize_face_terms && bdr_face_restrict_lex)
    {
-      auto restFbdr = dynamic_cast<const L2FaceRestriction&>(*bdr_face_restrict_lex);
-      restFbdr.AddFaceMatricesToElementMatrices(ea_data_bdr, ea_data);
+      auto restFbdr = dynamic_cast<const L2FaceRestriction*>(bdr_face_restrict_lex);
+      restFbdr->AddFaceMatricesToElementMatrices(ea_data_bdr, ea_data);
```
Comment (pazner) [inline fem/bilinearform_ext.cpp:574]: 👍

**Review by pazner — COMMENTED** [review]
Nice addition! Some minor comments…

File: tests/unit/fem/test_pa_kernels.cpp (hunk 3 above)
Comment (YohannDudouit) [reply tests/unit/fem/test_pa_kernels.cpp]: done in [45ac641](https://github.com/mfem/mfem/pull/2840/commits/45ac64129e74798ea90c9964717adfbe267bcb8c)

Comment (YohannDudouit) [reply tests/unit/fem/test_pa_kernels.cpp]: I can't believe I missed it right after explaining it...

File: fem/bilinearform_ext.cpp (hunk 4 above)
Comment (YohannDudouit) [reply fem/bilinearform_ext.cpp]: done in [a84b74d](https://github.com/mfem/mfem/pull/2840/commits/a84b74d625e17ac6ecc336921d982daf06bf2e9a)

File: fem/bilinearform_ext.cpp (hunk 5 above)
Comment (YohannDudouit) [reply fem/bilinearform_ext.cpp]: done in [a84b74d](https://github.com/mfem/mfem/pull/2840/commits/a84b74d625e17ac6ecc336921d982daf06bf2e9a)

**Comment by YohannDudouit** [comment]
@pazner I think I have addressed all your comments.

File: fem/bilinearform_ext.cpp [hunk 8]
```diff
@@ -903,6 +910,9 @@ void FABilinearFormExtension::Assemble()
             static_cast<const L2ElementRestriction*>(elem_restrict);
          const L2FaceRestriction *restF =
             static_cast<const L2FaceRestriction*>(int_face_restrict_lex);
+         MFEM_VERIFY(
+            !dynamic_cast<const NCL2FaceRestriction*>(int_face_restrict_lex),
+            "Full Assembly not yet supported on NCMesh.");
```
Comment (pazner) [inline fem/bilinearform_ext.cpp:915]: One more…

```suggestion
         MFEM_VERIFY(
            fes.Conforming(),
            "Full Assembly not yet supported on NCMesh.");
```

Comment (YohannDudouit) [reply fem/bilinearform_ext.cpp]: done in [4e77bad](https://github.com/mfem/mfem/pull/2840/commits/4e77bad9dd715d0fabd25bd34ed56f949a8079f8)

**Review by pazner — APPROVED** [review]
Awesome, looks great 😄

File: fem/bilinearform.cpp [hunk 9]
```diff
@@ -1064,6 +1064,19 @@ void BilinearForm::Mult(const Vector &x, Vector &y) const
...
 
+void BilinearForm::MultTranspose(const Vector & x, Vector & y) const
+{
+   if (ext)
+   {
+      ext->MultTranspose(x, y);
+   }
+   else
+   {
+      y = 0.0;
```
Comment (camierjs) [inline fem/bilinearform.cpp:1075]: Shouldn't this initialization be shared for both if-else-statements?
I see also that all `PAConvectionApplyT` but the `Smem` `3D` use `+=`: is it intended?

Comment (YohannDudouit) [reply fem/bilinearform.cpp:1075]: Good catch! This indeed was a bug, it also was in `SmemPAConvectionApply3D`.

Though, `BilinearFormExtension` only has `Mult` and `MultTranspose` (not `AddMult` and `AddMultTranspose`), so I think this is correct.

**Comment by YohannDudouit** [comment]
@pazner Apparently `trial_fes->Conforming()` broke something, I'm investigating.

**Comment by YohannDudouit** [comment]
@pazner @tzanio @artv3 I just realized that the support for `NCMesh` was not only working for partial assembly, but also for element assembly, I made a small modification to make sure it works with `NCMesh`. Should I mention it in `CHANGELOG`? 

Though @artv3 you have to be careful if you use the EA matrices on `NCMesh`, currently the `NCL2FaceRestriction` still contains the "nonconforming" contributions for the faces matrices.

**Comment by tzanio** [comment]
Re-merged in `next` for testing...

**Comment by tzanio** [comment]
@pazner and @artv3 -- do you want to review again?

**Review by pazner — APPROVED** [review]

**Review by artv3 — APPROVED** [review]

**Pull request merged**